### PR TITLE
HDDS-10200. OM may terminate due to NPE in `S3SecretValue` proto conversion

### PR DESCRIPTION
**Pull request by adoroszlai**
## What changes were proposed in this pull request?

Revoked S3 secret is stored in the cache as `S3SecretValue` with `null` secret value.  Since the object is mutable, other thread could access the same instance before revocation, and might encounter NPE while trying to convert to protobuf.

```
2024-01-24 09:49:09,063 [om1-OMDoubleBufferFlushThread] ERROR ratis.OzoneManagerDoubleBuffer (ExitUtils.java:terminate(133)) - Terminating with exit status 2: During flush to DB encountered error in OMDoubleBuffer flush thread om1-OMDoubleBufferFlushThread
org.apache.ratis.util.ExitUtils$ExitException: During flush to DB encountered error in OMDoubleBuffer flush thread om1-OMDoubleBufferFlushThread when handling OMRequest: cmdType: SetS3Secret
traceID: ""
success: true
status: OK
SetS3SecretResponse {
  accessId: "testUser1accessId1"
  secretKey: "testsecret"
}

	at org.apache.ratis.util.ExitUtils.terminate(ExitUtils.java:141)
	at org.apache.ratis.util.ExitUtils.terminate(ExitUtils.java:151)
	at org.apache.hadoop.ozone.om.ratis.OzoneManagerDoubleBuffer.terminate(OzoneManagerDoubleBuffer.java:559)
	at org.apache.hadoop.ozone.om.ratis.OzoneManagerDoubleBuffer.addToBatch(OzoneManagerDoubleBuffer.java:413)
	at org.apache.hadoop.ozone.om.ratis.OzoneManagerDoubleBuffer.flushBatch(OzoneManagerDoubleBuffer.java:345)
	at org.apache.hadoop.ozone.om.ratis.OzoneManagerDoubleBuffer.flushCurrentBuffer(OzoneManagerDoubleBuffer.java:319)
	at org.apache.hadoop.ozone.om.ratis.OzoneManagerDoubleBuffer.flushTransactions(OzoneManagerDoubleBuffer.java:286)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.NullPointerException
	at org.apache.hadoop.ozone.protocol.proto.OzoneManagerProtocolProtos$S3Secret$Builder.setAwsSecret(OzoneManagerProtocolProtos.java)
	at org.apache.hadoop.ozone.om.helpers.S3SecretValue.getProtobuf(S3SecretValue.java:100)
	at org.apache.hadoop.hdds.utils.db.DelegatedCodec.toCodecBuffer(DelegatedCodec.java:85)
	at org.apache.hadoop.hdds.utils.db.Codec.toDirectCodecBuffer(Codec.java:73)
	at org.apache.hadoop.hdds.utils.db.TypedTable.putWithBatch(TypedTable.java:171)
	at org.apache.hadoop.ozone.om.OmMetadataManagerImpl$1.addWithBatch(OmMetadataManagerImpl.java:1949)
	at org.apache.hadoop.ozone.om.response.s3.security.OMSetSecretResponse.addToDBBatch(OMSetSecretResponse.java:80)
	at org.apache.hadoop.ozone.om.response.OMClientResponse.checkAndUpdateDB(OMClientResponse.java:66)
	at org.apache.hadoop.ozone.om.ratis.OzoneManagerDoubleBuffer.lambda$9(OzoneManagerDoubleBuffer.java:407)
	at org.apache.hadoop.ozone.om.ratis.OzoneManagerDoubleBuffer.addToBatchWithTrace(OzoneManagerDoubleBuffer.java:244)
	at org.apache.hadoop.ozone.om.ratis.OzoneManagerDoubleBuffer.addToBatch(OzoneManagerDoubleBuffer.java:406)
	... 4 more
```

was seen in `TestMultiTenantVolume` (4 / 1000 runs).

This PR changes `S3SecretValue` to be immutable.

https://issues.apache.org/jira/browse/HDDS-10200

## How was this patch tested?

`TestMultiTenantVolume` passed 10x100:
https://github.com/adoroszlai/ozone/actions/runs/7645630287

Regular CI:
https://github.com/adoroszlai/ozone/actions/runs/7645622076

### CODE REVIEW

**Comment by adoroszlai** [comment]
@ArafatKhan2198 can you please also take a look?

**Review by sadanand48 — APPROVED** [review]
Thanks @adoroszlai  the patch, LGTM

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/S3InMemoryCache.java [hunk 1]
```diff
@@ -45,12 +45,9 @@ public void put(String id, S3SecretValue secretValue) {
...
   public void invalidate(String id) {
     S3SecretValue secret = cache.getIfPresent(id);
-    if (secret == null) {
-      return;
+    if (secret != null) {
+      cache.put(id, secret.deleted());
     }
-    secret.setDeleted(true);
-    secret.setAwsSecret(null);
-    cache.put(id, secret);
```
Comment (szetszwo) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/S3InMemoryCache.java:53]: Use `computeIfPresent`:
```java
    cache.asMap().computeIfPresent(id, (k, secret) -> secret.deleted());
```

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/S3SecretValue.java [hunk 2]
```diff
@@ -38,16 +38,29 @@ public static Codec<S3SecretValue> getCodec() {
   }
 
   // TODO: This field should be renamed to accessId for generalization.
-  private String kerberosID;
-  private String awsSecret;
-  private boolean isDeleted;
-  private long transactionLogIndex;
+  private final String kerberosID;
+  private final String awsSecret;
```
Comment (szetszwo) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/S3SecretValue.java:42]: An unrelated comment: It is not a good idea store a secret in a plaintext `String` and write it to db.

The usual practice is to encrypt it in `byte[]` and erase it after use.

File: hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/S3SecretValue.java [hunk 3]
```diff
@@ -38,16 +38,29 @@ public static Codec<S3SecretValue> getCodec() {
...
+    return new S3SecretValue(
+        Objects.requireNonNull(kerberosID),
+        Objects.requireNonNull(awsSecret),
+        false,
+        transactionLogIndex
+    );
+  }
+
+  public S3SecretValue deleted() {
+    return new S3SecretValue(kerberosID, awsSecret, true, transactionLogIndex);
```
Comment (szetszwo) [inline hadoop-ozone/common/src/main/java/org/apache/hadoop/ozone/om/helpers/S3SecretValue.java:60]: Pass an empty string for `awsSecret`?

**Review by szetszwo — COMMENTED** [review]
@adoroszlai , thanks for working on this!  Just a few minor comments inlined.

**Review by szetszwo — APPROVED** [review]
+1 the change looks good.

**Comment by adoroszlai** [comment]
Thanks @myskov, @sadanand48, @szetszwo for the review.

**Pull request merged**